free-2.md

```diff
@@ -1,69 +1,168 @@
+
+
+
 ## The user and a language
+
 This section describes who the project would serve and why a language might be a
+
 good way to meet their needs.
 
 
+
+This project would be for people who are trying to learn functional programming, either for their first exposure to CS or for someone who's used to imperative programming.
+
+
+
 ### What's the need?
+
 _What need is met by your idea? Who are you helping? What is that person's
+
 experience like now? What would their experience be like if you could help 
+
 them?_
 
 
+
+Imperative programming seems to be more commonly used for entry-level compter programmers, and while it has its benifits this makes people more hesitant or have more difficulty with undertstanding functional programming, which in some cases can be a more effective tool for the problem they're trying to solve. In addition, some people are more visual learners, and having a view of what's happening in a program would make debugging easier. With a learning tool, picking up a more foreign language would be easier and faster, and new programmers might start out with a deeper understanding of computer programming.
+
+
+
 ### Why a language?
+
 _Why is a DSL appropriate for your user(s)? How does it address the need?_
 
 
+
+Interacting directly with a programming language is one of the best ways to learn. So long as it's designed in a way that won't require unlearning later on, and allows for deep understanding of how the language operates, a learning language can be a powerful tool.
+
+
+
 ### Why you?
+
 _What excites you about this idea? How did you come up with it?_
 
 
+
+I've always been interested in teaching/tutoring how people learn, and while I'm not the absolute worst at functional programming languages, learning imperative languages was much easier for me.
+
+
+
 ### Domain
+
 _Describe the project's domain in five words._
 
 
+
+Visualizing general purpose functional languages ?
+
+
+
 ### Interface (syntax)
+
 _How might the user interact with the language? What does programming look 
+
 like? Why is this the right way to interact with the problem domain?_ 
 
 
+
+A program would just be lines of code as per usual.
+
+
+
 ### Operation (semantics)
+
 _What might happen when a program runs? How does a program interact with the
+
 user? What kinds of errors might occur, and how might they be communicated to
+
 the user?_
 
 
+
+The program would update the image when a button is clicked or when the code is executed. If there is an error, it would print the appropriate error message (in whatever style the host language uses), and some problems would be visually represented. For instance, if a function takes in two parameters but only one was given, the user would be able to see that a "piece" is missing; if object types are represented with tabs with the name of the type, then type mismatches would be easy to see.
```

Comment:
Mm, I'm very curious how one would flesh out this visual representation (especially with recursion and stuff like that).